File: Customer Data Analysis and Insights Project/scripts/r/eda.R
# Exploratory Data Analysis Script
# Customer Data Analysis

# Load libraries
library(tidyverse)
library(ggplot2)
library(dplyr)
library(VIM)

# Set paths
data_path <- "../../data/Customers.csv"
results_path <- "../../results"

# Create results directory if it doesn't exist
if (!dir.exists(results_path)) {
  dir.create(results_path, recursive = TRUE)
}

# Load data
df <- read.csv(data_path, stringsAsFactors = FALSE)

cat("==================================================\n")
cat("EXPLORATORY DATA ANALYSIS\n")
cat("==================================================\n\n")

# 1. Data Overview
cat("=== Dataset Information ===\n")
cat("Dataset loaded successfully!\n")
cat("Shape:", nrow(df), "rows,", ncol(df), "columns\n\n")
cat("First few rows:\n")
print(head(df))

cat("\n=== Dataset Structure ===\n")
str(df)

cat("\n=== Column Names ===\n")
print(colnames(df))

# 2. Missing Values Analysis
cat("\n=== Missing Values Analysis ===\n")
missing_values <- colSums(is.na(df))
missing_percentage <- (missing_values / nrow(df)) * 100

missing_df <- data.frame(
  Column = names(missing_values),
  Missing_Count = missing_values,
  Missing_Percentage = missing_percentage
)

missing_df_filtered <- missing_df[missing_df$Missing_Count > 0, ]
if (nrow(missing_df_filtered) == 0) {
  cat("✓ No missing values found in the dataset!\n")
} else {
  print(missing_df_filtered)
}

# 3. Duplicate Records
cat("\n=== Duplicate Records ===\n")
duplicate_count <- sum(duplicated(df))
cat("Number of duplicate records:", duplicate_count, "\n")
if (duplicate_count > 0) {
  cat("Duplicate records:\n")
  print(df[duplicated(df), ])
} else {
  cat("✓ No duplicate records found!\n")
}

# 4. Summary Statistics
cat("\n=== Numerical Columns Summary ===\n")
numerical_cols <- sapply(df, is.numeric)
if (sum(numerical_cols) > 0) {
  print(summary(df[, numerical_cols, drop = FALSE]))
} else {
  cat("No numerical columns found.\n")
}

cat("\n=== Categorical Columns Summary ===\n")
categorical_cols <- sapply(df, is.character) | sapply(df, is.factor)
for (col in names(df)[categorical_cols]) {
  cat("\n", col, ":\n", sep = "")
  cat("  Unique values:", n_distinct(df[[col]]), "\n")
  cat("  Most frequent:", names(sort(table(df[[col]]), decreasing = TRUE)[1]), "\n")
  cat("  Frequency:", max(table(df[[col]])), "\n")
}

# 5. Distribution Analysis - Customer ID
cat("\n=== Distribution Analysis ===\n")
if ("CustomerID" %in% colnames(df)) {
  png(file = file.path(results_path, "customer_id_distribution.png"), 
      width = 1200, height = 600, res = 300)
  hist(df$CustomerID, breaks = 20, main = "Distribution of Customer IDs",
       xlab = "Customer ID", ylab = "Frequency", col = "steelblue", border = "black")
  dev.off()
  cat("✓ Customer ID distribution plot saved\n")
}

# 6. Country Distribution
if ("Country" %in% colnames(df)) {
  country_counts <- df %>%
    count(Country) %>%
    arrange(desc(n))
  
  cat("\n=== Country Distribution ===\n")
  print(country_counts)
  
  png(file = file.path(results_path, "country_distribution.png"), 
      width = 1400, height = 800, res = 300)
  ggplot(country_counts, aes(x = reorder(Country, n), y = n)) +
    geom_bar(stat = "identity", fill = "steelblue", color = "black") +
    coord_flip() +
    labs(x = "Country", y = "Number of Customers", 
         title = "Customer Distribution by Country") +
    theme_minimal() +
    theme(axis.text.y = element_text(size = 8))
  dev.off()
  cat("✓ Country distribution plot saved\n")
}

# 7. City Distribution
if ("City" %in% colnames(df)) {
  city_counts <- df %>%
    count(City) %>%
    arrange(desc(n)) %>%
    head(20)
  
  cat("\n=== Top 20 Cities by Customer Count ===\n")
  print(city_counts)
  
  png(file = file.path(results_path, "city_distribution.png"), 
      width = 1400, height = 800, res = 300)
  ggplot(city_counts, aes(x = reorder(City, n), y = n)) +
    geom_bar(stat = "identity", fill = "coral", color = "black") +
    coord_flip() +
    labs(x = "City", y = "Number of Customers", 
         title = "Top 20 Cities by Customer Count") +
    theme_minimal() +
    theme(axis.text.y = element_text(size = 8))
  dev.off()
  cat("✓ City distribution plot saved\n")
}

# 8. Data Quality Checks
cat("\n=== Data Quality Checks ===\n")
cat("Unique Customer IDs:", n_distinct(df$CustomerID), "\n")
cat("Total Records:", nrow(df), "\n")
if (n_distinct(df$CustomerID) == nrow(df)) {
  cat("✓ All Customer IDs are unique\n")
} else {
  cat("⚠ Warning: Duplicate Customer IDs found\n")
}

# 9. Summary
cat("\n==================================================\n")
cat("EDA SUMMARY\n")
cat("==================================================\n")
cat("\n1. Dataset contains", nrow(df), "customers with", ncol(df), "attributes\n")
cat("2. Countries represented:", n_distinct(df$Country), "\n")
cat("3. Cities represented:", n_distinct(df$City), "\n")
cat("4. Missing values:", sum(is.na(df)), "\n")
cat("5. Duplicate records:", sum(duplicated(df)), "\n")

if ("Country" %in% colnames(df)) {
  top_country <- names(sort(table(df$Country), decreasing = TRUE)[1])
  top_country_count <- max(table(df$Country))
  cat("6. Most common country:", top_country, "(", top_country_count, "customers)\n")
}

if ("City" %in% colnames(df)) {
  top_city <- names(sort(table(df$City), decreasing = TRUE)[1])
  top_city_count <- max(table(df$City))
  cat("7. Most common city:", top_city, "(", top_city_count, "customers)\n")
}

cat("\n=== Key Insights ===\n")
cat("• The dataset appears to be clean with no missing values\n")
cat("• All customer IDs are unique\n")
cat("• The dataset has good geographical diversity\n")
cat("• Ready for further statistical and ML analysis\n")

cat("\n✓ EDA completed successfully!\n")

File: Customer Data Analysis and Insights Project/scripts/python/eda.py
"""
Exploratory Data Analysis Script
Customer Data Analysis
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from pathlib import Path

warnings.filterwarnings('ignore')

# Set style
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (12, 6)

# Set paths
data_path = Path('../../data/Customers.csv')
results_path = Path('../../results')
results_path.mkdir(exist_ok=True)

def load_data():
    """Load the dataset"""
    df = pd.read_csv(data_path)
    print("Dataset loaded successfully!")
    print(f"Shape: {df.shape}")
    return df

def data_overview(df):
    """Provide data overview"""
    print("\n=== Dataset Information ===")
    df.info()
    print(f"\n=== Dataset Shape ===")
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    print(f"\n=== Column Names ===")
    print(df.columns.tolist())
    return df.head()

def missing_values_analysis(df):
    """Analyze missing values"""
    missing_values = df.isnull().sum()
    missing_percentage = (missing_values / len(df)) * 100
    
    missing_df = pd.DataFrame({
        'Column': missing_values.index,
        'Missing Count': missing_values.values,
        'Missing Percentage': missing_percentage.values
    })
    
    print("\n=== Missing Values Analysis ===")
    result_df = missing_df[missing_df['Missing Count'] > 0]
    if len(result_df) == 0:
        print("✓ No missing values found in the dataset!")
    else:
        print(result_df)
    return missing_df

def duplicate_analysis(df):
    """Analyze duplicate records"""
    duplicate_count = df.duplicated().sum()
    print(f"\n=== Duplicate Records ===")
    print(f"Number of duplicate records: {duplicate_count}")
    if duplicate_count > 0:
        print("\nDuplicate records:")
        print(df[df.duplicated()])
    else:
        print("✓ No duplicate records found!")
    return duplicate_count

def summary_statistics(df):
    """Generate summary statistics"""
    print("\n=== Numerical Columns Summary ===")
    print(df.describe())
    
    print("\n=== Categorical Columns Summary ===")
    categorical_cols = df.select_dtypes(include=['object']).columns
    for col in categorical_cols:
        print(f"\n{col}:")
        print(f"  Unique values: {df[col].nunique()}")
        print(f"  Most frequent: {df[col].mode()[0] if len(df[col].mode()) > 0 else 'N/A'}")

def visualize_distributions(df):
    """Create distribution visualizations"""
    # Customer ID distribution
    plt.figure(figsize=(12, 6))
    plt.hist(df['CustomerID'], bins=20, edgecolor='black', alpha=0.7)
    plt.xlabel('Customer ID')
    plt.ylabel('Frequency')
    plt.title('Distribution of Customer IDs')
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(results_path / 'customer_id_distribution.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Country distribution
    country_counts = df['Country'].value_counts()
    plt.figure(figsize=(14, 8))
    country_counts.plot(kind='bar', color='steelblue', edgecolor='black')
    plt.xlabel('Country')
    plt.ylabel('Number of Customers')
    plt.title('Customer Distribution by Country')
    plt.xticks(rotation=45, ha='right')
    plt.grid(axis='y', alpha=0.3)
    plt.tight_layout()
    plt.savefig(results_path / 'country_distribution.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # City distribution (top 20)
    city_counts = df['City'].value_counts().head(20)
    plt.figure(figsize=(14, 8))
    city_counts.plot(kind='barh', color='coral', edgecolor='black')
    plt.xlabel('Number of Customers')
    plt.ylabel('City')
    plt.title('Top 20 Cities by Customer Count')
    plt.grid(axis='x', alpha=0.3)
    plt.tight_layout()
    plt.savefig(results_path / 'city_distribution.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    print("\n✓ Visualizations saved to results/ directory")

def data_quality_checks(df):
    """Perform data quality checks"""
    print("\n=== Data Quality Checks ===")
    
    # Check for empty strings
    print("\n=== Empty String Check ===")
    for col in df.columns:
        empty_count = (df[col] == '').sum()
        if empty_count > 0:
            print(f"{col}: {empty_count} empty strings")
    
    # Check data consistency
    print("\n=== Data Consistency Check ===")
    print(f"Unique Customer IDs: {df['CustomerID'].nunique()}")
    print(f"Total Records: {len(df)}")
    if df['CustomerID'].nunique() == len(df):
        print("✓ All Customer IDs are unique")
    else:
        print("⚠ Warning: Duplicate Customer IDs found")

def main():
    """Main function"""
    print("=" * 50)
    print("EXPLORATORY DATA ANALYSIS")
    print("=" * 50)
    
    # Load data
    df = load_data()
    
    # Data overview
    data_overview(df)
    
    # Missing values
    missing_values_analysis(df)
    
    # Duplicates
    duplicate_analysis(df)
    
    # Summary statistics
    summary_statistics(df)
    
    # Visualizations
    visualize_distributions(df)
    
    # Data quality checks
    data_quality_checks(df)
    
    # Summary
    print("\n" + "=" * 50)
    print("EDA SUMMARY")
    print("=" * 50)
    print(f"\n1. Dataset contains {df.shape[0]} customers with {df.shape[1]} attributes")
    print(f"2. Countries represented: {df['Country'].nunique()}")
    print(f"3. Cities represented: {df['City'].nunique()}")
    print(f"4. Missing values: {df.isnull().sum().sum()}")
    print(f"5. Duplicate records: {df.duplicated().sum()}")
    print(f"\n6. Most common country: {df['Country'].mode()[0]} ({df['Country'].value_counts().iloc[0]} customers)")
    print(f"7. Most common city: {df['City'].mode()[0]} ({df['City'].value_counts().iloc[0]} customers)")
    
    print("\n=== Key Insights ===")
    print("• The dataset appears to be clean with no missing values")
    print("• All customer IDs are unique")
    print("• The dataset has good geographical diversity")
    print("• Ready for further statistical and ML analysis")

if __name__ == "__main__":
    main()

